test-scenarios/engines-test.ts

```diff
@@ -0,0 +1,243 @@
+          export const nonce = Math.floor(Math.random() * 1000000);
+        `,
+      },
+    });
+
+    // if we let the app and the engine both directly auto-import "inner",
+    // ember-auto-import's own deduplication logic would ensure there's no
+    // duplication of code or state.
+    //
+    // But that only covers the top-level dependencies. All deeper dependencies
+    // are managed directly by webpack, and we also need those dependencies not
+    // to get duplicated module state.
+    //
+    // So this test gives the app and engine different first-level dependencies
+    // (outer1 and outer2) that both share a deeper dependency (inner).
+    for (let outer of ['outer1', 'outer2']) {
+      let dep = project.addDevDependency(outer, {
+        files: {
+          'index.js': `
+          export { nonce } from 'inner';
+        `,
+        },
+      });
+      dep.pkg.peerDependencies = { inner: '*' };
+      project.addDevDependency(dep);
+    }
+
+    merge(project.files, {
+      app: {
+        'router.js': `import EmberRouter from '@ember/routing/router';
+import config from './config/environment';
+
+const Router = EmberRouter.extend({
+  location: config.locationType,
+  rootURL: config.rootURL,
+});
+
+Router.map(function() {
+  this.mount('lazy-in-repo-engine', { path: '/use-lazy-engine', as: 'use-lazy-engine' });
+});
+
+export default Router;`,
+        templates: {
+          'index.hbs': '<div data-test-nonce>{{this.model.nonce}}</div>',
+        },
+        routes: {
+          'index.js': `import Route from '@ember/routing/route';
+          import { nonce } from 'outer1';
+
+          export default class IndexRoute extends Route {
+            model() {
+              return { nonce };
+            }
+          }`,
+        },
+      },
+      tests: {
+        acceptance: {
+          'basic-test.js': `import { module, test } from 'qunit';
+import { visit } from '@ember/test-helpers';
+import { setupApplicationTest } from 'ember-qunit';
+
+module('Acceptance | basics', function (hooks) {
+  setupApplicationTest(hooks);
+
+  // To control module caching behaviors, we intentionally only have a single test here that
+  // exercises everything we want to exercise. If we added more tests, their order could effect
+  // the outcome of when lazy modules are visible.
+
+  test('lazy engine basics', async function (assert) {
+
+    // 1. importing a plain npm pacakge from a lazy engines does not add the package eagerly
+
+    assert.equal(requirejs.entries['fake-package'], undefined, 'fake-package should not be loaded before visting lazy engine');
+    await visit('/use-lazy-engine');
+    assert.equal(typeof requirejs.entries['fake-package'], 'object', 'fake-package was loaded only after visting the engine');
+    assert
+      .dom('[data-test-pkg-name]')
+      .hasText('fake-package', 'The fake-package was correctly imported');
+
+    // 2. Deps that are shared between app and engine share the same module state
```

Comment:
We could write several more tests like this one that expose similar problems that remain un-addressed.

For example, if a dependency is used by two lazy engines but not the app, I think we will get the same duplicated module state problem.